SCons/Tool/MSCommon/vc.py

```diff
@@ -512,12 +512,10 @@ def find_batch_file(env,msvc_version,host_arch,target_arch):
     msvc_ver_numeric = get_msvc_version_numeric(msvc_version)
     use_arg = True
     vernum = float(msvc_ver_numeric)
-    if 7 <= vernum < 8:
-        pdir = os.path.join(pdir, os.pardir, "Common7", "Tools")
-        batfilename = os.path.join(pdir, "vsvars32.bat")
-    elif vernum < 7:
+    if vernum < 8:
```

Comment:
So the problem here was that it should also include 6.0 and above? Should we limit that way? or does this apply to even older MSVCs?(which are likely not in the "wild" anymore)

Reply:
***It may be helpful to understand the evolution of the msvc batch files in historical context for versions 6.0 to 8.0.  This may inform the current scons implementation for these versions.  It is difficult 20 years later to identify the exact intent but a compelling argument can be made nonetheless.***

The msvc ```vcvars32.bat``` batch files for versions 6.0 to 7.1 only target ```x86``` and require/accept no options.

The msvc ```vcvarsall.bat``` batch file for 8.0 introduced multiple build targets: ```x86```, ```amd64```,  ```ia64```, ```x86_amd64```, and ```x86_ia64```.  This is akin to the modern msvc batch files.

The msvc 6.0 ```vcvars32.bat``` batch file does not require or use any system environment variables specific to the msvc 6.0 installation.

The msvc 7.0 ```vcvars32.bat``` introduced an environment variable set by the installer that was the location of the common tools folder: ```VSCOMNTOOLS```.  The ```vcvars32.bat``` simply forwarded to the ```vsvars32.bat``` in the common folder via the environment ```VSCOMNTOOLS``` variable: ```%VSCOMNTOOLS%vsvars32.bat```.  Command-line users of msvc 7.0 would call ```vcvars32.bat``` and the batch files would automatically redirect the call to another batch file.  The ```vsvars32.bat``` in the common folder does not require or use any system environment variables specific to the msvc 7.0 installation.  By using ```<vcdir>\..\Common7\Tools\vsvars32.bat``` directly, there was no need for scons to import the common tools variable ```VSCOMNTOOLS``` from the environment.

The msvc 7.1 ```vcvars32.bat``` introduced an environment variable set by the installer that was the location of the common tools folder: ```VS71COMNTOOLS```.  The ```vcvars32.bat``` simply forwarded to the ```vsvars32.bat``` in the common folder via the environment ```VS71COMNTOOLS``` variable: ```%VS71COMNTOOLS%vsvars32.bat```.  Command-line users of msvc 7.1 would call ```vcvars32.bat``` and the batch files would automatically redirect the call to another batch file.  The ```vsvars32.bat``` in the common folder does not require or use any system environment variables specific to the msvc 7.1 installation.  By using ```<vcdir>\..\Common7\Tools\vsvars32.bat``` directly, there was no need for scons to import the common tools variable ```VS71COMNTOOLS``` from the environment.

The msvc 8.0 ```vcvars80.bat``` introduced an environment variable set by the installer that was the location of the common tools folder:  ```VS80COMNTOOLS```.  However, with the introduction of multiple build targets, the usage was different.  The default ```x86``` ```<vcdir>\bin\vcvars32.bat``` build would forward in a similar fashion as earlier batch files via the ```VS80COMNTOOLS``` variable.  Remaining build targets would define effectively "standalone" batch files.  The implication of this implementation meant that using the default target batch file ```<vcdir>\bin\vcvars32.bat``` requires the variable ```VS80COMNTOOLS``` to work properly.

Summary thus far:
* The current scons implementation for msvc 6.0 simply called ```vcvars32.bat```.  There was no common tools variable defined by the installer in the system environment.
* The current scons implementation for msvc 7.x calls ```vsvars32.bat``` in the common tools folder directly bypassing the traditional ```vcvars32.bat```.  Msvc 7.0 introduced the concept of the common tools environment variable that is set in the system environment (```VSCOMNTOOLS```).  Msvc 7.1 extended the concept of the common tools environment variable by introducing the version number in the variable name (```VS71COMNTOOLS```).
* The current scons implementation for msvc 8.0 *requires* the system environment variable ```VS80COMNTOOLS``` for default ```x86``` builds.

Conjecture: sometime later, the common tools variables were "back-ported" to 7.1, 7.0, and 6.0 as ```VS71COMNTOOLS``` (correct), ```VS70COMNTOOLS``` (incorrect/nonexistent), and ```VS60COMNTOOLS``` (incorrent/nonexistent).

Calling the 7.1 batch file ```vcvars32.bat``` directly via ```MSVC_USE_SCRIPT``` works as expected as the correct environment variable (```VS71COMNTOOLS```) is imported from the system environment.  

Due to the incorrect variable name (```VS70COMNTOOLS``` versus ```VSCOMNTOOLS```) used for 7.0, calling the 7.0 batch file ```vcvars32.bat``` directly via ```MSVC_USE_SCRIPT``` fails as the requisite environment variable (```VSCOMNTOOLS```) is not imported from the system environment.  The imported variable (```VS70COMNTOOLS```) has no effect because the msvc batch files aren't using that name.

Calling the 6.0 batch file ```vcvars32.bat``` directly via ```MSVC_USE_SCRIPT``` works as expected as there is no common tools environment variable necessary.  The imported variable (```VS60COMNTOOLS```) has no effect because the msvc batch files aren't using that name.

Once the "correct" environment variables (```VSCOMNTOOLS``` and ```VS71COMNTOOLS```) for 7.x are imported from the environment there is no need to call the ```vsvars32.bat``` batch file in the common folder.  The msvc ```vcvars32.bat``` batch file can be called directly as was done in 6.0.  This is why the code collapses to before version 8 (7.1, 7.0, and 6.0) use ```vcvars32.bat``` and for versions 8 and later use ```vcvarsall.bat```.

As for msvc 6.0, the installer only adds an (optional) environment variable to the system environment: ```MSDevDir```. As msvc 6.0 does not actually require an external variable specific to the installation this is unnecessary for the msvc batch files.  The vs.py code defines a common tools environment variable for all visual studio instances.  Since ```VS60COMNTOOLS``` never existed, the use of the only variable optionally defined by the msvc 6.0 installer (```MSDevDir```) seems appropriate and could be useful to an end-user.  The alternative is to allow an empty string (or None) in vs.py and check that the common tools variable is not empty or none before adding the variable to the common tools variable list. 

In a nutshell:
* ```VS70COMNTOOLS``` and ```VS60COMNTOOLS``` as used by scons do not exist in the wild.  Possibly the result of back-porting functionality to older code.
* Using the *correct* environment variable name ```VSCOMNTOOLS``` for 7.0 eliminates the need to call the common tools batch file directly simplifying the handling for msvc versions prior to 8.  This unifies the code for 6.0, 7.0, and 7.1 and all three versions succeed when calling the msvc ```vcvars32.bat``` batch file directly via ```MSVC_USE_SCRIPT```.
* The variable name, if any, to use for msvc 6.0 is an executive decision: nothing or ```MSDevDir```.  A case can be made that there is marginal utility is using the installer defined variable name.
* Passing the build argument is suppressed as the 6.0-7.1 batch files do not expect or use arguments.  This is done so that the internal scons msvc cache *could* be more effective in light of ```MSVC_USE_SCRIPT``` usage.

*I'm not exactly sure how to distill all of this in a sentence or two for the CHANGES.txt and RELEASE.txt files.*

Reply:
@bdbaddog I pushed a new CHANGES.txt and RELEASE.txt.  Your comment now shows as outdated without a response from me.  Are the updated blurbs heading in the right direction?

Reply:
ok, as always, a thorough explanation!

If I understand correctly:

1. These fixes only affect when you use MSVC_USE_SCRIPT pointing at vcvars32.bat ?
2. We have some junk variables we're copying from the post script shell which can be dropped (VS70COMNTOOLS, VS60COMNTOOLS).  Though removing them has no effect? So just cleanup?
3. VSCOMNTOOLS is only for 7.0
4. VS71COMNTOOLS (and similar named) are for 7.1 and above.
5. We need to add the OS var for 6.0 to avoid quoting the PATH which breaks finding DLLs?

So of the above, and please correct me if I'm wrong here, only the OS var missing breaks anything?
And the rest is cleanup of bad logic?

Reply:
> These fixes only affect when you use MSVC_USE_SCRIPT pointing at vcvars32.bat ?

Not quite.  

The batch file for 7.0 and 7.1 during normal detection has changed from ```<vcdir>..\common7\tools\vsvars32.bat``` to ```<vcdir>bin\vcvars32.bat```.  The vcvars32.bat batch files make use of the common environment variables ```VSCOMNTOOLS``` for 7.0 and ```VS71COMNTOOLS``` for 7.1 to call the appropriate vsvars32.bat file.  The vcvars32.bat files for 7.0 and 7.1 are only one-liners calling their respective vsvars32.bat batch file.

The inclusion of ```VSCOMNTOOLS``` allows the vcvars32.bat batch file to work correctly and therefore is used for both normal detection and via ```MSVC_USE_SCRIPT```.

> We have some junk variables we're copying from the post script shell which can be dropped (VS70COMNTOOLS, VS60COMNTOOLS). Though removing them has no effect? So just cleanup?

Correct.

> VSCOMNTOOLS is only for 7.0

Correct.

> VS71COMNTOOLS (and similar named) are for 7.1 and above.

Correct.

> We need to add the OS var for 6.0 to avoid quoting the PATH which breaks finding DLLs?

Technically, no.  The following is in common.py:
```
        for path in path_list:
            # Do not add empty paths (when a var ends with ;)
            if path:
                # XXX: For some reason, VC98 .bat file adds "" around the PATH
                # values, and it screws up the environment later, so we strip
                # it.
                path = path.strip('"')
```

The "some reason" is that the OS variable is not defined and the msvc 6.0 batch file thinks it is building on a Windows 9x platform.

However, both the ```OS``` and ```windir``` environment variables are used in the msvc 6.0 vsvars32.bat batch file and the SDK 6.0 (msvc 8.0) , 7.0 (msvc 9.0), and 7.1 (msvc 9.0) SetEnv.cmd batch files.

Without the ```OS``` variable defined, the SDK 6.0, 7.0, and 7.1 SetEnv.cmd issue an innocuous warning about building on a Windows 9x platform.  An end-user never sees the warnings though.

Without the ```windir``` environment variable a partial system path is included in the path variable.

From my own work, there is no apparent detrimental effect when importing both ```OS``` and ```windir``` from the system environment for all versions of msvc.  I am including both of those in my own work to eliminate warnings in the older SDK batch files and to eliminate the quoted paths for msvc 6.0.

> So of the above, and please correct me if I'm wrong here, only the OS var missing breaks anything?

It's not really broken as there is special case logic defined in common.py to remove the quotes from path elements for msvc 6.0.  Inclusion of ```OS``` *could* mean that the special handling in common.py could be removed.  This would require further testing.

> And the rest is cleanup of bad logic?

With the inclusion of the ```VSCOMNTOOLS``` variable, there is no longer a need to special-case the handling for 7.x.  Not necessarily bad logic in light of the history, perhaps better characterized as simpler and more robust after the change.

If desired, I can add the ```OS``` and/or ```windir``` environment variables.

Reply:
So adding OS and windir should allow simplifying the logic a bit?
If so, then definitely go ahead, and if you do, add that you are propagating those shell variables in CHANGES.txt and RELEASE.txt.

Reply:
I went ahead an added OS and windir and updated CHANGES.txt and RELEASE.txt